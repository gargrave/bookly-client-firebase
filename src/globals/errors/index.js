// @flow
import { parseError, parseFbError } from './apiErrorParser';
import { validationErrors } from './validationErrors';

export {
  parseError,
  parseFbError,
  validationErrors,
};
