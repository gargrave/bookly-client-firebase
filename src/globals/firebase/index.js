import firebase from 'firebase';
import 'firebase/firestore';

import firebaseConfig from '../../secrets/firebaseConfig.js';

let initialized = false;
let db;

if (!initialized) {
  firebase.initializeApp(firebaseConfig);
  db = firebase.firestore();
  initialized = true;
}

function docToObject(doc) {
  return {
    id: doc.id,
    ...doc.data(),
  };
}

function timestamp() {
  return firebase.firestore.FieldValue.serverTimestamp();
}

export {
  db,
  docToObject,
  timestamp,
};
