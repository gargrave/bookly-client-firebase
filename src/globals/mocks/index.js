import { authorMocks } from './data/authorMocks';
import { bookBucketMocks } from './data/bookBucketMocks';
import { bookMocks } from './data/bookMocks';
import { profileMocks } from './data/profileMocks';
import { userMocks } from './data/userMocks';

export {
  authorMocks,
  bookBucketMocks,
  bookMocks,
  profileMocks,
  userMocks,
};
