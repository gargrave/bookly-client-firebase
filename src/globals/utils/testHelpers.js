import React from 'react';
import { shallow } from 'enzyme';

export class ComponentBuilder {
  constructor(component, defaultProps) {
    this.component = component;
    this.defaultProps = defaultProps;
  }

  shallowGetComponent(extraProps = {}) {
    const props = {
      ...this.defaultProps,
      ...extraProps,
    };
    return shallow(<this.component {...props} />);
  }
}
