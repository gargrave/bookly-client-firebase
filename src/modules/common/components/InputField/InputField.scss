@import '../../../../styles/_colors';

.error {
  color: $errorColor;
}

input.invalid {
  border-color: $errorColor;
}
