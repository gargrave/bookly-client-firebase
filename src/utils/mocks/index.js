import authorMocks from './data/authors';
import userMocks from './data/users';

export {
  authorMocks,
  userMocks,
};
