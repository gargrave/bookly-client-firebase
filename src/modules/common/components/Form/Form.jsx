// @flow
import React from 'react'
import { any, array, func, bool, string } from 'prop-types'

import { buildClass } from '../../../../utils/cssHelpers'

import Alert from '../Alert/Alert'
import Button from '../Button'
import ButtonRow from '../ButtonRow/ButtonRow'

import styles from './Form.css'

type Props = {
  cancelBtnText?: string,
  children?: any,
  classes?: string[],
  disabled?: boolean,
  onCancel?: Function,
  onSubmit: Function,
  submitBtnText?: string,
  submitDisabled?: boolean,
  topLevelError?: string,
}

const Form = ({
  cancelBtnText,
  children,
  classes = [],
  disabled,
  onCancel,
  onSubmit,
  submitBtnText,
  submitDisabled = false,
  topLevelError,
}: Props) => {
  return (
    <div>
      {topLevelError && <Alert message={topLevelError} type="danger" />}
      <form
        className={buildClass(styles.form, ...(classes || []))}
        onSubmit={onSubmit}
        noValidate
      >
        {children}

        <ButtonRow>
          <Button
            canSubmit={true}
            disabled={submitDisabled || disabled || false}
            onClick={onSubmit}
            position="left"
            text={submitBtnText || 'Submit'}
            type="success"
          />

          {onCancel && (
            <Button
              onClick={onCancel}
              position="right"
              text={cancelBtnText || 'Cancel'}
              type="light"
            />
          )}
        </ButtonRow>
      </form>
    </div>
  )
}

Form.propTypes = {
  cancelBtnText: string,
  children: any,
  classes: array,
  disabled: bool,
  onCancel: func,
  onSubmit: func.isRequired,
  submitBtnText: string,
  submitDisabled: bool,
  topLevelError: string,
}

export default Form
