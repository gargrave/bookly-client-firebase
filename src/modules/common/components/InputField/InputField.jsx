// @flow
import React from 'react'
import { bool, func, number, oneOf, string } from 'prop-types'
import styled from 'react-emotion'

import { colors } from '../../../../styles'
import { clamp } from '../../../../utils/mathHelpers'

type Props = {
  boundValue: string,
  disabled?: boolean,
  error?: string,
  label?: string,
  maxLength?: number,
  name: string,
  onInputChange: Function,
  placeholder?: string,
  type?: string,
}

const acceptableTypes = ['email', 'password', 'search', 'text']

const StyledInput = styled('input')`
  border-color: ${props => props.invalid && colors.error} !important;
`

const StyledError = styled('p')`
  color: ${colors.error};
`

const InputField = ({
  boundValue,
  disabled,
  error,
  label,
  maxLength = 255,
  name,
  onInputChange,
  placeholder,
  type,
}: Props) => {
  return (
    <div className="input-field">
      {label && <label htmlFor={name}>{label}:</label>}

      <StyledInput
        disabled={disabled || false}
        id={name}
        invalid={!!error}
        maxLength={clamp(maxLength, 1, 255)}
        name={name}
        onChange={onInputChange}
        placeholder={placeholder || ''}
        type={type || 'text'}
        value={boundValue}
      />

      {error && <StyledError className="error">{error}</StyledError>}
    </div>
  )
}

InputField.propTypes = {
  boundValue: string.isRequired,
  disabled: bool,
  error: string,
  maxLength: number,
  label: string,
  name: string.isRequired,
  onInputChange: func.isRequired,
  placeholder: string,
  type: oneOf(acceptableTypes),
}

export default InputField
