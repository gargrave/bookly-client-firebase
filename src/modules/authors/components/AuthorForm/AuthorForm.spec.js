import { ComponentBuilder } from '../../../../utils/testHelpers'

import { authorMocks } from '../../../../globals/mocks/'

import Form from '../../../common/components/Form/Form'
import InputField from '../../../common/components/InputField/InputField'

import AuthorForm from './AuthorForm'

const defaultProps = {
  author: { ...authorMocks[0] },
  disabled: false,
  errors: {
    firstName: '',
    lastName: '',
  },
  onCancel: jest.fn(),
  onInputChange: jest.fn(),
  onSubmit: jest.fn(),
  submitDisabled: false,
  topLevelError: '',
}

const builder = new ComponentBuilder(AuthorForm, defaultProps)

describe('AuthorForm', () => {
  let component

  describe('with "author" populated', () => {
    it('matches the snapshot', () => {
      component = builder.shallowGetComponent()
      expect(component).toMatchSnapshot()
    })

    it('renders correctly', () => {
      component = builder.shallowGetComponent()
      expect(component.find(Form)).toHaveLength(1)
      expect(component.find(InputField)).toHaveLength(2)
    })
  })
})
