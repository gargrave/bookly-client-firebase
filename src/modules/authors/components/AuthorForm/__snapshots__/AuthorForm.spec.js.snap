// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AuthorForm with "author" populated matches the snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <AuthorForm
    author={
      Object {
        "bookCount": 1,
        "created": 2017-07-01T01:16:43.881Z,
        "firstName": "Aldous",
        "id": "oiw223h3qndlksadffl3ksdf",
        "lastName": "Huxley",
        "updated": 2017-07-01T01:16:43.881Z,
      }
    }
    disabled={false}
    errors={
      Object {
        "firstName": "",
        "lastName": "",
      }
    }
    onCancel={[Function]}
    onInputChange={[Function]}
    onSubmit={[Function]}
    submitDisabled={false}
    topLevelError=""
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": Array [
        <InputField
          boundValue="Aldous"
          disabled={false}
          error=""
          label="First name"
          name="firstName"
          onInputChange={[Function]}
        />,
        <InputField
          boundValue="Huxley"
          disabled={false}
          error=""
          label="Last name"
          name="lastName"
          onInputChange={[Function]}
        />,
      ],
      "classes": Array [
        "author-form",
      ],
      "disabled": false,
      "onCancel": [Function],
      "onSubmit": [Function],
      "submitDisabled": false,
      "topLevelError": "",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "boundValue": "Aldous",
          "disabled": false,
          "error": "",
          "label": "First name",
          "name": "firstName",
          "onInputChange": [Function],
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "boundValue": "Huxley",
          "disabled": false,
          "error": "",
          "label": "Last name",
          "name": "lastName",
          "onInputChange": [Function],
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": Array [
          <InputField
            boundValue="Aldous"
            disabled={false}
            error=""
            label="First name"
            name="firstName"
            onInputChange={[Function]}
          />,
          <InputField
            boundValue="Huxley"
            disabled={false}
            error=""
            label="Last name"
            name="lastName"
            onInputChange={[Function]}
          />,
        ],
        "classes": Array [
          "author-form",
        ],
        "disabled": false,
        "onCancel": [Function],
        "onSubmit": [Function],
        "submitDisabled": false,
        "topLevelError": "",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "boundValue": "Aldous",
            "disabled": false,
            "error": "",
            "label": "First name",
            "name": "firstName",
            "onInputChange": [Function],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "boundValue": "Huxley",
            "disabled": false,
            "error": "",
            "label": "Last name",
            "name": "lastName",
            "onInputChange": [Function],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    "attachTo": undefined,
    "hydrateIn": undefined,
  },
}
`;
